beacon-chain/rpc/helpers/broadcast.go

```diff
@@ -0,0 +1,173 @@
+package helpers
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/OffchainLabs/prysm/v6/beacon-chain/core/peerdas"
+	"github.com/OffchainLabs/prysm/v6/consensus-types/blocks"
+	ethpb "github.com/OffchainLabs/prysm/v6/proto/prysm/v1alpha1"
+	"github.com/pkg/errors"
+	"golang.org/x/sync/errgroup"
+)
+
+// DataColumnBroadcastFunc is a function for broadcasting data column sidecars.
+type DataColumnBroadcastFunc func(root [32]byte, subnet uint64, sidecar *ethpb.DataColumnSidecar) error
+
+// BlobBroadcastFunc is a function for broadcasting blob sidecars.
+type BlobBroadcastFunc func(ctx context.Context, index uint64, sidecar *ethpb.BlobSidecar) error
+
+// DataColumnOption configures options for broadcasting data column sidecars.
+type DataColumnOption func(*dataColumnOptions)
+
+type dataColumnOptions struct {
+	onReceiveDataColumns   func(columns []blocks.VerifiedRODataColumn) error
+	onDataColumnsProcessed func(columns []blocks.VerifiedRODataColumn)
+}
+
+// WithDataColumnReceiver sets the callback for receiving data columns.
+func WithDataColumnReceiver(fn func(columns []blocks.VerifiedRODataColumn) error) DataColumnOption {
+	return func(o *dataColumnOptions) {
+		o.onReceiveDataColumns = fn
+	}
+}
+
+// WithDataColumnProcessedCallback sets the callback to be called after data columns are processed.
+func WithDataColumnProcessedCallback(fn func(columns []blocks.VerifiedRODataColumn)) DataColumnOption {
+	return func(o *dataColumnOptions) {
+		o.onDataColumnsProcessed = fn
+	}
+}
+
+// BroadcastDataColumnSidecars broadcasts data column sidecars concurrently and optionally receives them.
+func BroadcastDataColumnSidecars(
+	ctx context.Context,
+	sidecars []*ethpb.DataColumnSidecar,
+	root [32]byte,
+	broadcastFunc DataColumnBroadcastFunc,
+	options ...DataColumnOption,
+) error {
+	opts := &dataColumnOptions{}
+	for _, opt := range options {
+		opt(opts)
+	}
+
+	verifiedRODataColumns := make([]blocks.VerifiedRODataColumn, 0, len(sidecars))
+
+	// Create verified RO data columns
+	for _, sd := range sidecars {
+		roDataColumn, err := blocks.NewRODataColumnWithRoot(sd, root)
+		if err != nil {
+			return errors.Wrap(err, "new read-only data column with root")
+		}
+
+		// We build this block ourselves, so we can upgrade the read only data column sidecar into a verified one.
+		verifiedRODataColumn := blocks.NewVerifiedRODataColumn(roDataColumn)
+		verifiedRODataColumns = append(verifiedRODataColumns, verifiedRODataColumn)
+	}
+
+	// Broadcast data columns concurrently
+	eg, _ := errgroup.WithContext(ctx)
+	for _, sd := range sidecars {
+		// Copy the iteration instance to a local variable to give each go-routine its own copy to play with.
+		// See https://golang.org/doc/faq#closures_and_goroutines for more details.
```

Comment:
Is this precaution still useful since go 1.22? https://go.dev/blog/loopvar-preview